book/chapters/03_data_prep.py

```diff
@@ -0,0 +1,161 @@
+# and through the [NSIDC website](https://nsidc.org/data/icesat-2/tools).
+# Here we will access data in the cloud by getting the appropriate s3urls using [icepyx](https://icepyx.readthedocs.io/en/latest/), a Python software library and community of ICESat-2 data users, developers, and data managers.
+# %% [markdown]
+# ## Getting started
+#
+# These are the tools you’ll need.
+
+# %%
+import datatree
+import earthaccess
+import icepyx as ipx
+import xarray as xr
+
+# %% [markdown] user_expressions=[]
+# Just to make sure we’re on the same page,
+# let’s check that we’ve got compatible versions installed.
+
+# %%
+xr.show_versions()
+
+
+# %% [markdown] user_expressions=[]
+# ## Cloud access to ICESat-2 ATL11 files
+#
+# In this book we use [icepyx](https://icepyx.readthedocs.io/en/latest/) for gathering the necessary s3 urls to access ICESat-2 data in the cloud.
+# icepyx can also be used (with nearly identical syntax) to download data to your local machine.
+# For more details on accessing ICESat-2 data in the cloud, please check out the references below!
+#
+# References:
+# - https://github.com/icesat2py/icepyx/blob/development/doc/source/example_notebooks/IS2_cloud_data_access.ipynb
+# - https://nsidc.github.io/earthaccess/tutorials/demo/
+# - https://nasa-openscapes.github.io/earthdata-cloud-cookbook/examples/NSIDC/ICESat2-CMR-AWS-S3.html#data-access-using-aws-s3
+# - https://nsidc.org/data/user-resources/help-center/nasa-earthdata-cloud-data-access-guide
+# - https://book.cryointhecloud.com/tutorials/IS2_ATL15_surface_height_anomalies/IS2_ATL15_surface_height_anomalies.html
+
+# %% [markdown] user_expressions=[]
+# ### Providing credentials
+#
+# Accessing NASA data requires you to have an Earthdata Login.
+# You can sign up for one free at https://data.nsidc.earthdatacloud.nasa.gov/
+# and learn more about NASA authentication and managing your credentials [via Earthaccess](https://nsidc.github.io/earthaccess/) and
+# [here](https://nasa-openscapes.github.io/2021-Cloud-Hackathon/tutorials/04_NASA_Earthdata_Authentication.html#authentication-for-nasa-earthdata).
+#
+# By obtaining your s3 urls via icepyx, you are also able to authenticate for cloud data access (note: Earthaccess is used under the hood to do this).
+# %%
+# First we must let icepyx know where (and when) we would like data from.
+
+short_name = "ATL11"  # The data product we would like to query
+spatial_extent = [-180.0, -85.0, 180.0, -60.0]  # bounding box for Antarctica
+date_range = ["2018-09-15", "2023-05-31"]  # entire satellite record
+# %%
+# Setup the Query object
+region = ipx.Query(short_name, spatial_extent, date_range)
+# %%
+region.visualize_spatial_extent()
+
+# %%
+# Get the granule IDs and cloud access urls (note that due to some missing ICESat-2 product metadata, icepyx is still working to provide s3 urls for some products)
+gran_ids = region.avail_granules(ids=True, cloud=True)
+print(len(gran_ids[0]))
+print(gran_ids[0][:10])
```

Comment:
@JessicaS11, I'm seeing this warning when printing the granule IDs

```
/srv/conda/envs/notebook/lib/python3.10/site-packages/icepyx/core/granules.py:86: UserWarning: We are still working in implementing ID generation for this data product.
  warnings.warn("We are still working in implementing ID generation for this data product.", UserWarning)
```
Assuming that the PR at https://github.com/icesat2py/icepyx/pull/426 fixes this somewhat? Will the full s3 urls (e.g. `s3://nsidc-cumulus-prod-protected/ATLAS/ATL11/005/2019/09/30/ATL11_005411_0315_005_03.h5`) be returned in the list, or just the HDF5 filename (e.g. `ATL11_005411_0315_005_03.h5`)? I'm hoping for the full s3 urls (because it's been a pain working out how the ATL11 files are organized in the S3 bucket) :pray:

Reply:
> I'm seeing this warning when getting the granule IDs

Are you using the latest dev version? I think I only noted it in slack, but

> Assuming that the PR at https://github.com/icesat2py/icepyx/pull/426 fixes this somewhat?

Is correct and should fix it entirely (but hasn't been pushed through to a new release yet; I thought I took the warning off but perhaps not?). It returns the full s3 urls (agree on the pain point!), and the rest of this workflow (including reading in with datatree) worked successfully for me. If you're working in CryoCloud, you may need to jump through some extra hoops to use the dev install (so I've made a habit of checking my version at every import).

Reply:
Ah cool, let me try things out on the dev branch, and yes I was working on the CryoCloud :smiley:

Reply:
Yep, using `pip install git+https://github.com/icesat2py/icepyx.git@development` to get `icepyx=0.7.1.dev9+g66e2863` returns the full s3 urls! I can't tell you how happy I am not having to manually generate a list of 4000+ urls in a file like this [`ATL11_to_download.txt`](https://github.com/weiji14/deepicedrain/blob/v0.4.2/ATL11_to_download.txt) anymore :smiley:

Gonna play with those ATL11 files and xarray-datatree a bit more now, hoping that this [`map_over_subtree`](https://xarray-datatree.readthedocs.io/en/latest/generated/datatree.map_over_subtree.html) method can save me from writing some nasty for-loops.